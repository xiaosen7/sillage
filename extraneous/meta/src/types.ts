import type { CSSProperties } from "react";

// /////////////////////////////////////////////////////////////////////////////
// #region Node
// /////////////////////////////////////////////////////////////////////////////

interface NodeBase {
  id: string;
  name: string;
  children: JsonNode[];
  passProps: {
    style?: CSSProperties;
    [K: string]: any;
  };
  position: [number, number];
}

/**
 * only used by container
 * @defaultValue "free"
 */
export type LayoutType = "column" | "row" | "free";

interface NodeType extends NodeBase {
  isContainer: false;
}

interface ContainerNodeType extends NodeBase {
  isContainer: true;
  layoutType: LayoutType;
}

export type JsonNode = ContainerNodeType | NodeType;

export interface JsonRootNode extends ContainerNodeType {
  name: "root";
  position: [0, 0];
  layoutType: "free";
}

// /////////////////////////////////////////////////////////////////////////////
// #endregion
// /////////////////////////////////////////////////////////////////////////////

// /////////////////////////////////////////////////////////////////////////////
// #region Meta
// /////////////////////////////////////////////////////////////////////////////

export interface ComponentMetaConfig
  extends Pick<JsonNode, "name" | "isContainer"> {
  isInternal: boolean;
  layoutType?: LayoutType;
}

// /////////////////////////////////////////////////////////////////////////////
// #endregion
// /////////////////////////////////////////////////////////////////////////////
