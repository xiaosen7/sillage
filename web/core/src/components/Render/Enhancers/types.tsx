import { type ReactElement } from "react";
import { type Node } from "../../../models";

export interface EnhancerProps {
  children: ReactElement;
  node: Node;
  [K: string]: any;
}
