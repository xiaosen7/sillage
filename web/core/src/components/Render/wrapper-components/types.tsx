import { type ReactElement } from "react";
import { type Node } from "../../../models";

export interface WrapperComponentProps {
  children: ReactElement;
  node: Node;
  [K: string]: any;
}
