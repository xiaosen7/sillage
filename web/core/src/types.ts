import { Bridge } from "./models";
import type { ComponentType } from "react";

export * from "@sillage/meta";
export type EditorComponentType<T = any> = ComponentType<{
  onChange: (value: T) => void;
  value: T;
}>;
