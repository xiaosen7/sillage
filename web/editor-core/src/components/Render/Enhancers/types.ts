import { type Node } from "@sillage/core";
import { type ReactElement } from "react";

export interface EnhancerProps {
  node: Node;
  children: ReactElement;
  innerWrapperProps?: any;
}
