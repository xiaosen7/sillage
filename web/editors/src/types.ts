import type { ComponentType } from "react";

export type EditorComponentType<T = any> = ComponentType<{
  onChange: (value: T) => void;
  value: T;
}>;
