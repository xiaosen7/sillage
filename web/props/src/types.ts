import type { Bridge } from "@sillage/core";
import type { CSSProperties } from "react";

export interface EditorProps {
  value?: any;
  onChange: (value: any) => void;
}

export type MaterialComponentProps<T> = {
  bridge: Bridge;
  style?: CSSProperties;
} & T;
