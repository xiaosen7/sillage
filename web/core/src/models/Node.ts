import { Map, Set, is } from "immutable";
import { Emitter, genId } from "@sillage/utils";
import { type CSSProperties } from "react";
import { type JsonNode, type LayoutType } from "../types";
import { type Material } from "./Materials";

enum Topic {
  PropsUpdate,
  PropUpdate,
  ChildChange,
  PositionUpdate,
  StylePropertyChange,
}

const PassProps = "passProps";
const Name = "name";
const IsContainer = "isContainer";
const Position = "position";
const Children = "children";
const Id = "id";
const Style = "style";
const LayoutTypeKey = "layoutType";

const Parent = "_parent";

/**
 * 所有的节点数据保存在这里
 */
export class Node extends Emitter<Topic> {
  static readonly Topic = Topic;
  private data: Map<string, any>;
  private mountPoint: HTMLElement | null = null;

  constructor(initialJsonNode: JsonNode) {
    super();

    const { children, ...rest } = initialJsonNode;

    this.data = Map(rest);

    // link children
    this.data = this.data.set(Children, Set());
    for (const jsonNode of children) {
      const node = new Node(jsonNode);
      this.linkChild(node);
    }
  }

  static fromMaterial(m: Material) {
    const { metaConfig, name } = m;
    const { initialProps } = m;

    const isContainer = !!metaConfig.isContainer;

    const jsonNode = {
      children: [] as JsonNode[],
      isContainer,
      name,
      position: [0, 0],
      passProps: initialProps,
      id: genId(),
      style: {},
      layoutType: isContainer ? metaConfig.layoutType ?? "free" : undefined,
    };

    return new Node(jsonNode as any as JsonNode);
  }

  getId(): string {
    return this.data.get(Id);
  }

  toJSON(): JsonNode {
    const object = this.data.toJS();
    Reflect.deleteProperty(object, Parent);
    const children = object.children as Node[];
    object.children = children.map((x) => x.toJSON());
    return object as JsonNode;
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #region structure
  // /////////////////////////////////////////////////////////////////////////////

  isRoot() {
    return this.getName() === "root";
  }

  getParent(): Node {
    return this.data.get(Parent);
  }

  setParent(parent?: Node) {
    this.data = this.data.set(Parent, parent);
  }

  hasChild(node: Node) {
    return this.getChildren().has(node);
  }

  /**
   * delete child and set `child.parent` to `undefined`
   * @param node
   */
  unlinkChild(child: Node) {
    this.deleteChild(child);
    child.setParent();
  }

  linkChild(child: Node) {
    if (this.hasChild(child) && child.getParent() === this) {
      return;
    }

    if (child.getParent()) {
      // break link
      const nodeParent = child.getParent();
      nodeParent.unlinkChild(child);
    }

    child.setParent(this);
    const children = this.getChildren();
    this.data = this.data.set(Children, children.add(child));
    this.emit(Node.Topic.ChildChange);
  }

  /**
   * only delete child in parent
   * @param node
   */
  deleteChild(node: Node) {
    const children = this.getChildren().delete(node);
    this.data = this.data.set(Children, children);
    this.emit(Node.Topic.ChildChange);
  }

  getChildren(): Set<Node> {
    return this.data.get(Children);
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #endregion
  // /////////////////////////////////////////////////////////////////////////////

  // /////////////////////////////////////////////////////////////////////////////
  // #region props
  // /////////////////////////////////////////////////////////////////////////////

  setPassProp(prop: string, value: any) {
    this.data = this.data.setIn([PassProps, prop], value);
    this.emit(Topic.PropUpdate, prop);
  }

  getPassProp(prop: string) {
    return this.data.getIn([PassProps, prop]);
  }

  getPassProps() {
    return this.data.get(PassProps);
  }

  setPassProps(passProps: any) {
    this.data = this.data.set(PassProps, passProps);
    this.emit(Topic.PropsUpdate, passProps);
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #endregion
  // /////////////////////////////////////////////////////////////////////////////

  // /////////////////////////////////////////////////////////////////////////////

  // /////////////////////////////////////////////////////////////////////////////
  // #region style
  // /////////////////////////////////////////////////////////////////////////////

  getStyle(): CSSProperties {
    return this.data.getIn([PassProps, Style]) as CSSProperties;
  }

  setStyleProperty(property: keyof CSSProperties, value: string | number) {
    console.log("setStyleProperty", property, value);
    this.data = this.data.setIn([PassProps, Style, property], value);
    this.emit(Topic.StylePropertyChange, property);
  }

  getStyleProperty(property: keyof CSSProperties): number | string {
    return this.data.getIn([PassProps, Style, property]) as number | string;
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #endregion
  // /////////////////////////////////////////////////////////////////////////////

  getName(): string {
    return this.data.get(Name);
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #region container property
  // /////////////////////////////////////////////////////////////////////////////

  isContainer(): boolean {
    return this.data.get(IsContainer);
  }

  getLayoutType(): LayoutType {
    return this.data.get(LayoutTypeKey);
  }

  // /////////////////////////////////////////////////////////////////////////////
  // #endregion
  // /////////////////////////////////////////////////////////////////////////////

  getPosition(): [number, number] {
    return this.data.get(Position);
  }

  setPosition(position: [number, number]) {
    if (is(position, this.getPosition())) {
      return;
    }

    this.data = this.data.set(Position, position);
    this.emit(Topic.PositionUpdate);
  }

  setMountPoint(el: HTMLElement) {
    if (this.mountPoint) {
      return false;
    }

    this.mountPoint = el;
    return true;
  }

  getMountBoundingRect() {
    return this.mountPoint?.getBoundingClientRect();
  }

  clone() {
    const json = this.toJSON();
    json.id = genId();
    return new Node(json);
  }
}
