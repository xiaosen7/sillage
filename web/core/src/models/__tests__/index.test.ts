import * as vt from "vitest";
import { type JsonNode, type JsonRootNode } from "../../types";
import { Node } from "../Node";

vt.describe("Node", () => {
  const json: JsonNode = {
    children: [],
    isContainer: false,
    name: "rect",
    passProps: {
      width: 390,
      height: 844,
    },
    position: [0, 0],
    id: "rectId",
  };

  vt.test("position", () => {
    const node = new Node(json);
    node.setPosition([20, 20]);
    vt.expect(node.getPosition()).toEqual([20, 20]);
    vt.expect(node.toJSON().position).toEqual([20, 20]);
  });

  vt.test("passProps", () => {
    const node = new Node(json);

    node.setPassProp("name", "rect");
    vt.expect(node.getPassProp("name")).toEqual("rect");
    vt.expect(node.getPassProps()).toEqual({
      width: 390,
      height: 844,
      name: "rect",
    });

    node.setPassProps({ name: "rect" });
    vt.expect(node.getPassProp("name")).toEqual("rect");
    vt.expect(node.toJSON().passProps).toEqual({
      name: "rect",
    });
  });

  vt.test("children", () => {
    const node = new Node(json);
    const child = new Node({
      children: [],
      isContainer: false,
      name: "button",
      passProps: {},
      position: [0, 0],
      id: "buttonId",
    });

    node.linkChild(child);
    vt.expect(node.hasChild(child)).toBeTruthy();
    vt.expect(node.toJSON()).toMatchSnapshot();
    vt.expect(child.getParent()).toBe(node);

    child.setPosition([30, 30]);
    vt.expect(node.toJSON()).toMatchSnapshot();

    node.deleteChild(child);
    vt.expect(node.hasChild(child)).toBeFalsy();
    vt.expect(child.getParent()).toBe(node);

    node.unlinkChild(child);
    vt.expect(child.getParent()).toBe(undefined);
  });

  vt.test("initialize children", () => {
    const defaultJsonNode: JsonRootNode = {
      children: [
        {
          name: "rect",
          isContainer: false,
          children: [],
          passProps: {
            width: 100,
            height: 100,
          },
          id: "rect",
          position: [0, 0],
        },
      ],
      isContainer: true,
      name: "root",
      passProps: {
        // iphone 12
        width: 390,
        height: 844,
      },
      position: [0, 0],
      id: "root",
      layoutType: "free",
    };

    const root = new Node(defaultJsonNode);
    const rect = root.getChildren().toJS()[0] as Node;

    vt.expect(rect.getName()).toBe("rect");
    vt.expect(rect.getParent()).toBe(root);
    vt.expect(root.hasChild(rect)).toBe(true);
  });
});
