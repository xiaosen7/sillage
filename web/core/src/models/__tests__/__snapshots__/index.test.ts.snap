// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Node > children 1`] = `
{
  "children": [
    {
      "children": [],
      "id": "buttonId",
      "isContainer": false,
      "name": "button",
      "passProps": {},
      "position": [
        0,
        0,
      ],
    },
  ],
  "id": "rectId",
  "isContainer": false,
  "name": "rect",
  "passProps": {
    "height": 844,
    "width": 390,
  },
  "position": [
    0,
    0,
  ],
}
`;

exports[`Node > children 2`] = `
{
  "children": [
    {
      "children": [],
      "id": "buttonId",
      "isContainer": false,
      "name": "button",
      "passProps": {},
      "position": [
        30,
        30,
      ],
    },
  ],
  "id": "rectId",
  "isContainer": false,
  "name": "rect",
  "passProps": {
    "height": 844,
    "width": 390,
  },
  "position": [
    0,
    0,
  ],
}
`;
